flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java

```diff
@@ -150,6 +172,103 @@ public void recordFailedCommit() {
+     */
+    public void maybeAddRecordsLagMetric(KafkaConsumer<?, ?> consumer, TopicPartition tp) {
+        // Lazily register pendingRecords
+        if (recordsLagMetrics == null) {
+            this.recordsLagMetrics = new ConcurrentHashMap<>();
+            this.sourceReaderMetricGroup.setPendingRecordsGauge(
+                    () -> {
+                        long pendingRecordsTotal = 0;
+                        for (Metric recordsLagMetric : this.recordsLagMetrics.values()) {
+                            pendingRecordsTotal +=
+                                    ((Double) recordsLagMetric.metricValue()).longValue();
+                        }
+                        return pendingRecordsTotal;
+                    });
+        }
+
+        recordsLagMetrics.computeIfAbsent(
+                tp,
+                (ignored) -> {
+                    try {
+                        return KafkaMetricUtils.getKafkaMetric(
+                                consumer.metrics(),
+                                CONSUMER_FETCH_MANAGER_GROUP,
+                                RECORDS_LAG,
+                                (metric) -> {
+                                    final Map<String, String> tags = metric.metricName().tags();
+                                    return tags.containsKey("topic")
+                                            && tags.get("topic").equals(tp.topic())
+                                            && tags.containsKey("partition")
+                                            && tags.get("partition")
+                                                    .equals(String.valueOf(tp.partition()));
+                                });
+                    } catch (IllegalStateException e) {
+                        LOG.warn(
+                                String.format(
+                                        "Error when getting Kafka consumer metric \"%s\" "
+                                                + "for partition \"%s\". "
+                                                + "Metric \"%s\" may not be reported correctly. ",
+                                        RECORDS_LAG, tp, MetricNames.PENDING_BYTES),
+                                e);
+                        return null;
+                    }
+                });
+    }
+
+    /**
+     * Remove a partition's records-lag metric from tracking list.
+     *
+     * @param tp Unassigned topic partition
+     */
+    public void removeRecordsLagMetric(TopicPartition tp) {
+        if (recordsLagMetrics != null) {
+            recordsLagMetrics.remove(tp);
+        }
+    }
+
+    /** Update {@link org.apache.flink.runtime.metrics.MetricNames#IO_NUM_BYTES_IN}. */
+    public void updateNumBytesInCounter() {
+        if (this.bytesConsumedTotalMetric != null) {
+            KafkaMetricUtils.sync(
+                    this.bytesConsumedTotalMetric,
+                    this.sourceReaderMetricGroup.getIOMetricGroup().getNumBytesInCounter());
```

Comment:
I think we also need sum up the metrics per split, right? Currently, if we have two splits and consumers, they would overwrite each other.

Reply:
I think ```bytes-consumed-total``` without topic in the tag already sums them up. Also there should be only one consumer instance running in one subtask, so there won't be overwrite I think

Reply:
Sorry I was confused. I thought there was a `SplitReader` per split but that's obviously not the case.